research/deep_speech/deep_speech_model.py

```diff
@@ -0,0 +1,174 @@
+# Copyright 2018 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Network structure for DeepSpeech model."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+from six.moves import xrange  # pylint: disable=redefined-builtin
+import tensorflow as tf
+
+_SUPPORTED_RNNS = {
+    "lstm": tf.keras.layers.LSTM,
+    "rnn": tf.keras.layers.SimpleRNN,
+    "gru": tf.keras.layers.GRU,
+}
+
+
+def _conv_bn_layer(cnn_input, filters, kernel_size, strides, layer_id):
+  """2D convolution + batch normalization layer
+  Args:
+    cnn_input: input data for convolution layer.
+    filters: an integer, number of output filters in the convolution.
+    kernel_size: a tuple specifying the height and width of the 2D convolution
+      window.
+    strides: a tuple specifying the stride length of the convolution.
+    layer_id: an integer specifying the layer index.
+
+  Returns:
+    tensor output from the current layer.
+  """
+  cnn_output = tf.keras.layers.Conv2D(
+      filters=filters, kernel_size=kernel_size, strides=strides, padding="same",
+      activation="linear", name="cnn_{}".format(layer_id))(cnn_input)
+  output = tf.keras.layers.BatchNormalization(
+      momentum=0.1, epsilon=1e-05)(cnn_output)
+  return output
+
+
+def _rnn_layer(input_data, rnn_cell, rnn_hidden_size, layer_id, rnn_activation,
+              is_batch_norm, is_bidirectional):
+  """Defines a batch normalization + rnn layer.
+  Args:
+    input_data: input tensors for the current layer.
+    rnn_cell: RNN cell instance to use.
+    rnn_hidden_size: an integer for the dimensionality of the rnn output space.
+    layer_id: an integer for the index of current layer.
+    rnn_activation: activation function to use.
+    is_batch_norm: a boolean specifying whether to perform batch normalization
+      on input states.
+    is_bidirectional: a boolean specifying whether the rnn layer is
+      bi-directional.
+
+  Returns:
+    tensor output for the current layer.
+  """
+  if is_batch_norm:
+    input_data = tf.keras.layers.BatchNormalization(
+        momentum=0.1, epsilon=1e-05)(input_data)
```

Comment:
Ah, I can answer my own question: these should be constants at the top of the module.